controllers/agents/app/serviceclaim_controller.go

```diff
@@ -43,12 +58,146 @@ type ServiceClaimReconciler struct {
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(&sclaim, ServiceClaimFinalizer) {
+			controllerutil.AddFinalizer(&sclaim, ServiceClaimFinalizer)
+			if err := r.Update(ctx, &sclaim); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(&sclaim, ServiceClaimFinalizer) {
+			// our finalizer is present, so lets handle any external dependency
+			sclaimCopy := r.createServiceClaimCopy(sclaim, deployment, remote_namespace)
+			if err := r.deleteExternalResources(ctx, sclaimCopy, remote_client); err != nil {
+				// if fail to delete the external dependency here, return with error
+				// so that it can be retried
+				return ctrl.Result{}, err
+			}
+
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(&sclaim, ServiceClaimFinalizer)
+			if err := r.Update(ctx, &sclaim); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
 
-	// TODO(user): your logic here
+	sclaimCopy := r.createServiceClaimCopy(sclaim, deployment, remote_namespace)
+	spec := sclaimCopy.Spec
+	op, err := controllerutil.CreateOrUpdate(ctx, remote_client, sclaimCopy, func() error {
+		sclaimCopy.Spec = spec
+		return nil
+	})
+	if err != nil {
+		l.Error(err, "Failed to create service claim",
+			"service", sclaim.Name,
+			"namespace", sclaim.Namespace)
+	} else {
+		l.Info("Wrote service claim", "claim", sclaim.Name, "namespace", sclaim.Namespace, "operation", op)
+	}
+
+	sclaim.Status.ClaimID = sclaimCopy.Status.ClaimID
+	sclaim.Status.State = sclaimCopy.Status.State
+	sclaim.Status.RegisteredService = sclaimCopy.Status.RegisteredService
+	if err := r.Status().Update(ctx, &sclaim); err != nil {
+		l.Error(err, "unable to update the ServiceClaim", "ServiceClaim", sclaim)
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
+func (r *ServiceClaimReconciler) createServiceClaimCopy(sclaim primazaiov1alpha1.ServiceClaim, deployment appsv1.Deployment, remote_namespace string) *primazaiov1alpha1.ServiceClaim {
+	sclaimCopy := sclaim.DeepCopy()
+	sclaimCopy.Spec.EnvironmentTag = ""
+	sclaimCopy.Spec.ApplicationClusterContext = &primazaiov1alpha1.ServiceClaimApplicationClusterContext{}
+	sclaimCopy.Spec.ApplicationClusterContext.ClusterEnvironmentName = deployment.Labels["primaza.io/cluster-environment"]
+	sclaimCopy.Spec.ApplicationClusterContext.Namespace = sclaim.Namespace
+	sclaimCopy.Namespace = remote_namespace
+	sclaimCopy.ResourceVersion = ""
+	return sclaimCopy
+}
+
+func (r *ServiceClaimReconciler) deleteExternalResources(ctx context.Context, sclaim *primazaiov1alpha1.ServiceClaim, cli client.Client) error {
+	if err := cli.Delete(ctx, sclaim); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ServiceClaimReconciler) GetPrimazaKubeconfig(ctx context.Context, namespace string) (*rest.Config, string, error) {
```

Comment:
This function can be removed now, and we can use the method https://github.com/primaza/primaza/blob/main/pkg/primaza/workercluster/connection.go#L73.

Reply:
Done